File: sercop-unified/frontend/src/services/cpBudgetService.ts
import { get, post, put } from '../utils/apiClient';

// ============================================================================
// Types
// ============================================================================

export interface CPBudgetCertificate {
  id: string;
  processId: string;
  paaItemId: string | null;
  certificateNumber: string;
  certificateDate: string;
  amount: number;
  budgetPartition: string;
  fundingSource: string;
  fiscalYear: number;
  status: string;
  erpReference: string | null;
  erpResponse: string | null;
  createdAt: string;
  updatedAt: string;
  createdBy: string;
}

export interface CPBudgetExecution {
  id: string;
  certificateId: string;
  executionType: string;
  amount: number;
  executionDate: string;
  documentNumber: string;
  createdAt: string;
}

// ============================================================================
// API Functions
// ============================================================================

const BASE_URL = '/api/compras-publicas/budget';

export const createCertificate = async (data: {
  processId: string;
  paaItemId?: string;
  amount: number;
  budgetPartition: string;
  fundingSource: string;
  fiscalYear: number;
}): Promise<CPBudgetCertificate> => {
  const response = await post(`${BASE_URL}/certificates`, data);
  return response.json();
};

export const getCertificate = async (id: string): Promise<CPBudgetCertificate> => {
  const response = await get(`${BASE_URL}/certificates/${id}`);
  return response.json();
};

export const getCertificatesByProcess = async (processId: string): Promise<CPBudgetCertificate[]> => {
  const response = await get(`${BASE_URL}/certificates/process/${processId}`);
  return response.json();
};

export const updateCertificateStatus = async (
  id: string,
  status: string
): Promise<CPBudgetCertificate> => {
  const response = await put(`${BASE_URL}/certificates/${id}/status`, { status });
  return response.json();
};

export const addExecution = async (
  certId: string,
  data: {
    executionType: string;
    amount: number;
    documentNumber: string;
  }
): Promise<CPBudgetExecution> => {
  const response = await post(`${BASE_URL}/certificates/${certId}/executions`, data);
  return response.json();
};

export const getExecutions = async (certId: string): Promise<CPBudgetExecution[]> => {
  const response = await get(`${BASE_URL}/certificates/${certId}/executions`);
  return response.json();
};

// ============================================================================
// Helpers
// ============================================================================

export const getCertificateStatusColor = (status: string): string => {
  const colors: Record<string, string> = {
    SOLICITADO: 'blue',
    APROBADO: 'green',
    BLOQUEADO: 'orange',
    LIBERADO: 'teal',
    CANCELADO: 'red',
  };
  return colors[status] || 'gray';
};

export const getExecutionTypeLabel = (type: string): string => {
  const labels: Record<string, string> = {
    COMPROMISO: 'Compromiso',
    DEVENGADO: 'Devengado',
    PAGO: 'Pago',
  };
  return labels[type] || type;
};

export const getExecutionTypeColor = (type: string): string => {
  const colors: Record<string, string> = {
    COMPROMISO: 'blue',
    DEVENGADO: 'orange',
    PAGO: 'green',
  };
  return colors[type] || 'gray';
};

File: sercop-unified/backend-java/src/main/java/com/globalcmx/api/compraspublicas/budget/service/CPBudgetService.java
package com.globalcmx.api.compraspublicas.budget.service;

import com.globalcmx.api.compraspublicas.budget.entity.CPBudgetCertificate;
import com.globalcmx.api.compraspublicas.budget.entity.CPBudgetExecution;
import com.globalcmx.api.compraspublicas.budget.repository.CPBudgetCertificateRepository;
import com.globalcmx.api.compraspublicas.budget.repository.CPBudgetExecutionRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class CPBudgetService {

    private final CPBudgetCertificateRepository budgetCertificateRepository;
    private final CPBudgetExecutionRepository budgetExecutionRepository;

    @Transactional("readModelTransactionManager")
    public CPBudgetCertificate createCertificate(String processId, String paaItemId, String certificateNumber,
                                                  LocalDate certificateDate, BigDecimal amount,
                                                  String budgetPartition, String fundingSource,
                                                  Integer fiscalYear, String userId) {
        log.info("Creating budget certificate: process={}, certificate={}, amount={}",
                processId, certificateNumber, amount);

        CPBudgetCertificate certificate = CPBudgetCertificate.builder()
                .id(UUID.randomUUID().toString())
                .processId(processId)
                .paaItemId(paaItemId)
                .certificateNumber(certificateNumber)
                .certificateDate(certificateDate)
                .amount(amount)
                .budgetPartition(budgetPartition)
                .fundingSource(fundingSource)
                .fiscalYear(fiscalYear)
                .status("SOLICITADO")
                .createdBy(userId)
                .createdAt(LocalDateTime.now())
                .build();

        return budgetCertificateRepository.save(certificate);
    }

    @Transactional(transactionManager = "readModelTransactionManager", readOnly = true)
    public CPBudgetCertificate getCertificate(String id) {
        return budgetCertificateRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Budget certificate not found: " + id));
    }

    @Transactional(transactionManager = "readModelTransactionManager", readOnly = true)
    public List<CPBudgetCertificate> getCertificatesByProcess(String processId) {
        return budgetCertificateRepository.findByProcessIdOrderByCreatedAtDesc(processId);
    }

    @Transactional("readModelTransactionManager")
    public CPBudgetCertificate updateCertificateStatus(String id, String status) {
        log.info("Updating budget certificate status: id={}, status={}", id, status);

        CPBudgetCertificate certificate = budgetCertificateRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Budget certificate not found: " + id));

        certificate.setStatus(status);
        return budgetCertificateRepository.save(certificate);
    }

    @Transactional("readModelTransactionManager")
    public CPBudgetExecution addExecution(String certificateId, String executionType, BigDecimal amount,
                                           LocalDate executionDate, String documentNumber) {
        log.info("Adding budget execution: certificate={}, type={}, amount={}",
                certificateId, executionType, amount);

        // Verify that the certificate exists
        budgetCertificateRepository.findById(certificateId)
                .orElseThrow(() -> new IllegalArgumentException("Budget certificate not found: " + certificateId));

        CPBudgetExecution execution = CPBudgetExecution.builder()
                .id(UUID.randomUUID().toString())
                .certificateId(certificateId)
                .executionType(executionType)
                .amount(amount)
                .executionDate(executionDate)
                .documentNumber(documentNumber)
                .createdAt(LocalDateTime.now())
                .build();

        return budgetExecutionRepository.save(execution);
    }

    @Transactional(transactionManager = "readModelTransactionManager", readOnly = true)
    public List<CPBudgetExecution> getExecutions(String certificateId) {
        return budgetExecutionRepository.findByCertificateIdOrderByExecutionDateAsc(certificateId);
    }
}
